pennylane/gradients/gradient_transform.py

```diff
@@ -0,0 +1,178 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This module contains utilities for defining custom gradient transforms,
+including a decorator for specifying gradient expansions."""
+# pylint: disable=too-few-public-methods
+import pennylane as qml
+
+
+unsupported_op = lambda op: op.grad_method is None
+supported_op = lambda op: op.grad_method is not None
+trainable_op = lambda op: any(qml.math.requires_grad(p) for p in op.parameters)
+
+
+def gradient_expand(tape, depth=10):
+    """Expand out a tape so that it supports differentiation
+    of requested operations.
+
+    This is achieved by decomposing all trainable operations that have
+    ``Operation.grad_method=None`` until all resulting operations
+    have a defined gradient method.
+
+    Args:
+        tape (.QuantumTape): the input tape to expand
+        depth (int) : the maximum expansion depth
+
+    Returns:
+        .QuantumTape: the expanded tape
+    """
+
+    # check if the tape contains unsupported trainable operations
+    if any(unsupported_op(op) and trainable_op(op) for op in tape.operations):
+
+        # Define the stopping condition for the expansion
+        stop_cond = lambda obj: (
+            not isinstance(obj, qml.measure.MeasurementProcess)
+            and ((supported_op(obj) and trainable_op(obj)) or not trainable_op(obj))
+        )
+
+        return tape.expand(depth=depth, stop_at=stop_cond)
+
+    return tape
+
+
+class gradient_transform(qml.batch_transform):
+    """Decorator for defining quantum gradient transforms.
+
+    Quantum gradient transforms are a specific case of :class:`~.batch_transform`.
+    All quantum gradient transforms accept a tape, and outputs
+    a batch of tapes to be independently executed on a quantum device, alongside
+    a post-processing function that returns the result.
+
+    Args:
+        expand_fn (function): An expansion function (if required) to be applied to the
+            input tape before the gradient computation takes place. If not provided,
+            the default expansion function simply expands all operations that
+            have ``Operation.grad_method=None`` until all resulting operations
+            have a defined gradient method.
+        differentiable (bool): Specifies whether the gradient transform is differentiable or
+            not. A transform may be non-differentiable if it does not use an
+            autodiff framework for its tensor manipulations. In such a case, setting
+            ``differentiable=False`` instructs the decorator
+            to mark the output as 'constant', reducing potential overhead.
+        hybrid (bool): Specifies whether classical processing inside a QNode
+            should be taken into account when transforming a QNode.
+
+            - If ``True``, and classical processing is detected and this
+              option is set to ``True``, the Jacobian of the classical

```

Comment:
What is the classical processing used for in the case where it is present?

Reply:
This is a good question :)

Reply:
So!

Reply:
Recall that gradient transforms are of the form `tape -> processing_fn(execute(gradient_tapes))`. That is, they will return the Jacobian of the quantum circuit output with respect to *gate arguments*.

This is all fine and very well-defined when applying a gradient transform to a tape.

However, when applying a gradient transform to a _QNode_, there is a subtlety --- the arguments you pass to evlauate the transform are *QNode arguments*, which are not necessarily *gate arguments*! In fact, it is *very* easy to introduce classical processing in-between the QNode arguments and the gate arguments without realizing.

Consider the following:

```python
@qml.qnode(dev)
def circuit(weights):
    qml.RX(weights[1], wires=0)
    qml.RY(2 * weights[0] + weights[2], wires=0)
    return qml.probs(wires=0)
```

Here, we have added classical processing inbetween the QNode and gate arguments by permuting the arguments, and multiplying by a scalar. The processing function mapping `qnode args -> gate args` looks like this:

```
C: (w0, w1, w2) -> (w1, 2 w0 + w2)
```

If we were to evaluating the gradient transform ignoring this classical processing, we would extract only the quantum Jacobian. This would be a `(2, 2)` matrix (2 gate arguments, 2 output dimensions), and would not be what the user expects at all.

Instead, the user would expect a `(3, 2)` matrix output (3 QNode args, 2 output dimensions). So what we need to do is compute the ('classical') Jacobian of the `C` function above (which will be `(3, 2)`) and multiply it by the quantum Jacobian:

```
Jac = CJac @ QJac = (3, 2) @ (2, 2) = (3, 2)
```

Reply:
@josh146 thanks for the explanation, this helped me understand what is happening in the code below :100:

Reply:
No worries! Even _just_ permuting the gate arguments wrt qnode arguments is non-trivial classical processing - the classical Jacobian will be a permutation matrix, not an identity matrix!

Reply:
This is the part of the code I was most worried about, hence all the different classical processing tests 😆

Reply:
@glassnotes this is also similar to the issue you had with the Fourier module I believe? Except it is a bit harder to solve; simply knowing the Jacobian of the classical processing was insufficient, in that case, you needed the _actual_ gate mapping function `C`.

Something maybe we could do with tracing 😄

Reply:
Thanks for elaborating on this @josh146 , I've definitely used classical processing between the QNode arguments and gate arguments without  realizing! I'll be keeping an eye out for this now